platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/EmbeddedSolrFactory.java

```diff
@@ -46,167 +42,176 @@
       configStore.setDataDirectoryPath(System.getProperty("solr.data.dir"));
     }
-
-    ConfigurationFileProxy configProxy = new ConfigurationFileProxy(configStore);
-    return Futures.immediateFuture(
-        getEmbeddedSolrServer(
-            core, HttpSolrClientFactory.DEFAULT_SOLRCONFIG_XML, null, configProxy));
-  }
-
-  /**
-   * Creates a new {@link EmbeddedSolrServer} using the default Solr core name ({@value
-   * #DEFAULT_EMBEDDED_CORE_NAME}), configuration file ({@value
-   * HttpSolrClientFactory#DEFAULT_SOLRCONFIG_XML}) and schema.
-   *
-   * @return a new {@link EmbeddedSolrServer} instance
-   */
-  public static SolrClient getEmbeddedSolrServer() {
-    return getEmbeddedSolrServer(
-        DEFAULT_EMBEDDED_CORE_NAME, HttpSolrClientFactory.DEFAULT_SOLRCONFIG_XML, null, null);
+    return newClient(
+        core,
+        HttpSolrClientFactory.DEFAULT_SOLRCONFIG_XML,
+        HttpSolrClientFactory.DEFAULT_SCHEMA_XML,
+        configStore,
+        new ConfigurationFileProxy(configStore));
   }
 
   /**
-   * Creates a new {@link EmbeddedSolrServer} using the Solr configuration file provided. Uses the
-   * default core name ({@value #DEFAULT_EMBEDDED_CORE_NAME}) and schema.
+   * Requests the creation of a new embedded {@code SolrClient} for a specific Solr core name.
    *
-   * @param solrConfigXml name of the Solr configuration file. Defaults to {@value
-   *     HttpSolrClientFactory#DEFAULT_SOLRCONFIG_XML} if {@code null}.
-   * @return a new {@link EmbeddedSolrServer} instance
+   * @param core the name of the Solr core to create to create a client for
+   * @param configXml the name of the Solr configuration file
+   * @param schemaXml the name of the Solr core schema
+   * @param configStore the {@link ConfigurationStore} instance to use
+   * @param configProxy the {@link ConfigurationFileProxy} instance to use
+   * @return the newly created {@code SolrClient}
+   * @throws IllegalArgumentException if <code>core</code>, <code>configXml</code>, <code>schemaXml
+   * </code>, <code>configStore</code> or <code>configProxy</code> is <code>null</code>
    */
-  public static EmbeddedSolrServer getEmbeddedSolrServer(@Nullable String solrConfigXml) {
-    return getEmbeddedSolrServer(DEFAULT_EMBEDDED_CORE_NAME, solrConfigXml, null, null);
+  public org.codice.solr.client.solrj.SolrClient newClient(
+      String core,
+      String configXml,
+      String schemaXml,
+      ConfigurationStore configStore,
+      ConfigurationFileProxy configProxy) {
+    Validate.notNull(core, "invalid null Solr core name");
+    Validate.notNull(configXml, "invalid null Solr config file");
+    Validate.notNull(schemaXml, "invalid null Solr schema file");
+    Validate.notNull(configStore, "invalid null Solr config store");
+    Validate.notNull(configProxy, "invalid null Solr config proxy");
+    LOGGER.debug("Solr({}): Creating an embedded Solr client", core);
+    return new SolrClientAdapter(
```

Comment:
❓Could we inject a `Supplier<SolrClientAdapter>` to this class and delegate to it instead of doing the `new` here? This would allow the tests to inject a synchronous `SolrClientAdapter` and avoid potential timing and threading issues in the tests. We also wouldn't have to make the `getEmbeddedSolrServer()` package-private as it would get called right away.

Reply:
Scratch the `Supplier<>` option. Just realized that we're passing in parameters to the constructor. `BiFunction<>` would be a better option :-)

Reply:
I believe the timing and threading issues for the factory tests are being alleviated already in a very simple way by simply stubbing the creators. Agreed that it is still asynchronous but that part is quick and only requires one to wait. As such, the test cases and the code here can remain simple enough. I fear in having too many injection points where we end up testing something completely different than what will be used at runtime.